hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java

```diff
@@ -55,91 +55,62 @@ public UtilPrngHandler() {
         }
     }
 
-    /**
-     * Get the n-3 running hash bytes from the n-3 running hash. If the running hash is not
-     * available, will throw a HandleException. n-3 running hash is chosen for generating random
-     * number instead of n-1 running hash for processing transactions quickly. Because n-1 running
-     * hash might not be available when the transaction is processed and might need to wait longer.
-     *
-     * @param nMinus3RunningHash n-3 running hash
-     * @return n-3 running hash bytes
-     */
-    private byte[] getNMinus3RunningHashBytes(@NonNull final RunningHash nMinus3RunningHash) {
-        // This can't happen because this running hash is taken from record stream.
-        // If this happens then there is a bug in the code.
-        requireNonNull(nMinus3RunningHash);
-
-        Hash nMinusThreeHash;
-        try {
-            nMinusThreeHash = nMinus3RunningHash.getFutureHash().get();
-            // Use n-3 running hash instead of n-1 running hash for processing transactions quickly
-            if (nMinusThreeHash == null || Arrays.equals(nMinusThreeHash.getValue(), new byte[48])) {
-                log.info("No n-3 record running hash available to generate random number");
-                return MISSING_BYTES;
-            }
-            // generate binary string from the running hash of records
-            return nMinusThreeHash.getValue();
-        } catch (InterruptedException e) {
-            log.error("Interrupted exception while waiting for n-3 running hash", e);
-            Thread.currentThread().interrupt();
-            throw new HandleException(ResponseCodeEnum.UNKNOWN);
-            // FUTURE : Need to decide on the response code for this case
-        } catch (ExecutionException e) {
-            log.error("Unable to get current n-3 running hash", e);
-            throw new HandleException(ResponseCodeEnum.UNKNOWN);
-            // FUTURE : Need to decide on the response code for this case
-        }
-    }
-
     /**
      * Generate a random number from the given bytes in the given range
      * @param pseudoRandomBytes bytes to generate random number from
      * @param range range of the random number
      * @return random number
      */
-    private int randomNumFromBytes(@NonNull final byte[] pseudoRandomBytes, final int range) {
-        final int initialBitsValue = ByteBuffer.wrap(pseudoRandomBytes, 0, 4).getInt();
-        return IntMath.mod(initialBitsValue, range);
+    private int randomNumFromBytes(@NonNull final Bytes pseudoRandomBytes, final int range) {
+        // Use the initial 4 bytes of the random number to extract the integer value. This might be a negative number,
+        // which when used with the modulus operator will result in a negative number. To avoid this, we mask off the
+        // high bit to make sure we always have a positive number for the mod operator.
+        final int initialBitsValue = pseudoRandomBytes.getInt(0) & 0x7fffffff;
```

Comment:
This has some interesting interactions with twos-compliment, which might be worth noting (you're changing the number returned by just masking the bit).  Alternatively an absolute-value conversion (`Math.abs` or `value < 0 ? -value : value` style) does not change the magnitude of the number returned.

Reply:
Yes, that is right (and I'm impressed you found me out!). Since we're talking about random numbers here, switching from -1 to Integer.MAX_VALUE by masking off the high bit, for instance, should be fine, and will be exceptionally fast. I didn't see any reason to do more work to convert -7 to 7, for instance. And this is compatible with the logic we had before from the Google library.